### PR TITLE
API ESDT Supply

### PR DESCRIPTION
**Pull request by miiu96**
Added a new API route `/network/esdt/supply/:token` that will return the total supply of the provided token.

### CODE REVIEW

File: api/mock/facade.go [hunk 1]
```diff
@@ -192,6 +193,14 @@ func (f *Facade) GetEnableEpochsMetrics() (*data.GenericAPIResponse, error) {
 	return f.GetEnableEpochsMetricsHandler()
 }
 
+func (f *Facade) GetESDTSupply(token string) (*data.ESDTSupplyResponse, error) {
```
Comment (iulianpascalau) [inline api/mock/facade.go:196]: missing comment

File: process/esdtSuppliesProcessor.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+	networkESDTSupplyPath = "/network/esdt/supply/"
+)
+
+type esdtSuppliesProcessor struct {
+	baseProc    Processor
+	scQueryProc SCQueryService
+}
+
+func NewESDTSuppliesProcessor(baseProc Processor, scQueryProc SCQueryService) (*esdtSuppliesProcessor, error) {
```
Comment (iulianpascalau) [inline process/esdtSuppliesProcessor.go:24]: missing comments on exported items

File: process/esdtSuppliesProcessor.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	if baseProc == nil {
```
Comment (iulianpascalau) [inline process/esdtSuppliesProcessor.go:25]: check if nils?

File: process/mock/scQueryServiceStub.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package mock
+
+import (
+	"github.com/ElrondNetwork/elrond-go/data/vm"
+	"github.com/ElrondNetwork/elrond-proxy-go/data"
+)
+
+// SCQueryServiceStub is a stub
```
Comment (iulianpascalau) [inline process/mock/scQueryServiceStub.go:8]: or simply add comments like:
```GO
// SCQueryServiceStub -
```

File: api/mock/facade.go (hunk 1 above)
Comment (miiu96) [reply api/mock/facade.go:196]: added.

File: process/esdtSuppliesProcessor.go (hunk 2 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:23]: added

File: process/mock/scQueryServiceStub.go (hunk 4 above)
Comment (miiu96) [reply process/mock/scQueryServiceStub.go:8]: done.

File: process/esdtSuppliesProcessor.go (hunk 3 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:24]: done.

**Review by iulianpascalau — APPROVED** [review]

File: api/groups/baseNetworkGroup.go [hunk 5]
```diff
@@ -136,3 +139,25 @@ func (group *networkGroup) getEnableEpochs(c *gin.Context) {
...
 }
+
+func (group *networkGroup) getESDTSupply(c *gin.Context) {
+	tokenIdentifier := c.Param("token")
+	if tokenIdentifier == "" {
+		shared.RespondWith(
+			c,
+			http.StatusBadRequest,
+			nil,
+			fmt.Sprintf("%v: %v", errors.ErrGetESDTTokenData, errors.ErrEmptyTokenIdentifier),
```
Comment (bogdan-rosianu) [inline api/groups/baseNetworkGroup.go:150]: `fmt.Sprintf("%s: %s", errors.ErrGetESDTTokenData.Error(), errors.ErrEmptyTokenIdentifier.Error()),`

File: data/esdt.go [hunk 6]
```diff
@@ -9,6 +9,17 @@ const (
...
 var ValidTokenTypes = []string{FungibleTokens, SemiFungibleTokens, NonFungibleTokens}
 
+// ESDTSupplyResponse is a response holding esdt supply
+type ESDTSupplyResponse struct {
+	Data  ESDTSupply `json:"data"`
+	Error string     `json:"error"`
+	Code  ReturnCode `json:"code"`
+}
+
+type ESDTSupply struct {
```
Comment (bogdan-rosianu) [inline data/esdt.go:19]: add comment

File: process/esdtSuppliesProcessor.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+	totalSupply.Add(totalSupply, initialSupply)
+	res.Data.Supply = totalSupply.String()
+
+	return res, nil
+}
+
+func (esp *esdtSuppliesProcessor) getSupplyFromShards(tokenIdentifier string) (*big.Int, error) {
+	totalSupply := big.NewInt(0)
+	shardIDSs := esp.baseProc.GetShardIDs()
```
Comment (bogdan-rosianu) [inline process/esdtSuppliesProcessor.go:65]: shardIDs

File: process/esdtSuppliesProcessor.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(res.ReturnData) < 4 {
+		return big.NewInt(0), nil
+	}
+
+	supplyBytes := res.ReturnData[3]
+	supplyBig, _ := big.NewInt(0).SetString(string(supplyBytes), 10)
```
Comment (bogdan-rosianu) [inline process/esdtSuppliesProcessor.go:99]: don't ignore the boolean return param. better return big.NewInt(0) if parsing fails

File: process/esdtSuppliesProcessor.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		return nil, errObs
+	}
+
+	apiPath := networkESDTSupplyPath + token
+	for _, observer := range shardObservers {
+		var responseEsdtSupply data.ESDTSupplyResponse
+
+		_, errGet := esp.baseProc.CallGetRestEndPoint(observer.Address, apiPath, &responseEsdtSupply)
+		if errGet != nil {
+			log.Error("esdt supply request", "observer", observer.Address, "error", errGet.Error())
```
Comment (bogdan-rosianu) [inline process/esdtSuppliesProcessor.go:115]: also log the shard ID

File: process/esdtSuppliesProcessor.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+func (esp *esdtSuppliesProcessor) GetESDTSupply(tokenIdentifier string) (*data.ESDTSupplyResponse, error) {
+	totalSupply, err := esp.getSupplyFromShards(tokenIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &data.ESDTSupplyResponse{}
+	if !isFungibleESDT(tokenIdentifier) {
+		res.Data.Supply = totalSupply.String()
+		return res, nil
```
Comment (bogdan-rosianu) [inline process/esdtSuppliesProcessor.go:49]: the `code` field of ESDTSupplyResponse will remain unpopulated. please treat this

File: process/esdtSuppliesProcessor.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	initialSupply, err := esp.getInitialSupplyFromMeta(tokenIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	totalSupply.Add(totalSupply, initialSupply)
+	res.Data.Supply = totalSupply.String()
```
Comment (bogdan-rosianu) [inline process/esdtSuppliesProcessor.go:58]: the `code` field of ESDTSupplyResponse will remain unpopulated. please treat this

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: data/esdt.go (hunk 6 above)
Comment (miiu96) [reply data/esdt.go:19]: added.

File: process/esdtSuppliesProcessor.go (hunk 7 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:65]: renamed.

File: process/esdtSuppliesProcessor.go (hunk 8 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:99]: done.

File: process/esdtSuppliesProcessor.go (hunk 9 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:115]: added.

File: process/esdtSuppliesProcessor.go (hunk 10 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:49]: added. good catch.

File: process/esdtSuppliesProcessor.go (hunk 11 above)
Comment (miiu96) [reply process/esdtSuppliesProcessor.go:58]: added.

File: versions/factory/versionedFacadeCreator.go [hunk 12]
```diff
@@ -103,6 +104,7 @@ func createVersionV1_0Facade(facadeArgs FacadeArgs) (*facadeVersions.ElrondProxy
 		ValidatorStatisticsProcessor: facadeArgs.ValidatorStatisticsProcessor,
 		ProofProcessor:               facadeArgs.ProofProcessor,
 		PubKeyConverter:              facadeArgs.PubKeyConverter,
+		ESDTSuppliesProcessor:        facadeArgs.ESDTSuppliesProcessor,
```
[the remark this answers is not in the document]
Comment (miiu96) [reply versions/factory/versionedFacadeCreator.go:107]: renamed.

**Review by bogdan-rosianu — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**